### PR TITLE
Add a transitiveClosure() function

### PR DESCRIPTION
**Pull request by nex3**
This also makes topologicalSort() and stronglyConnectedComponents()
iterative rather than recursive, so transitiveClosure() is able to
operate on very large graphs.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,12 +1,11 @@
-# 2.2.1-dev
+# 2.3.0
 
-- Require Dart 2.18
+- Add a `transitiveClosure` function.
 
-# 2.2.0
```
Comment (jakemac53) [inline CHANGELOG.md:5]: It looks like this was accidentally removed

File: lib/src/strongly_connected_components.dart [hunk 2]
```diff
@@ -45,21 +45,45 @@ List<List<T>> stronglyConnectedComponents<T extends Object>(
...
-    onStack.add(node);
-    for (final next in edges(node)) {
+  final stack = [for (final node in nodes) _StackState(node)];
+  outer:
+  while (stack.isNotEmpty) {
+    final state = stack.removeLast();
+    final node = state.node;
+    var iterator = state.iterator;
+
+    late int lowLink;
```
Comment (jakemac53) [inline lib/src/strongly_connected_components.dart:55]: It looks like we could just leave this as nullable too - although having it late in a sense might be less bug prone as we would get an exception on line 86 if we don't ever assign a value. Ultimately either way is ok with me.

I do have a general preference for nullable versus late, with explicit `!` assertions. We could add one of those on line 86 as well, to retain the check that we did assign something, even though the compiler/analyzer wouldn't require it for a simple equality check.

File: test/transitive_closure_test.dart [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
+/// Returns the transitive closure of a graph represented a map from keys to
+/// edges.
+Map<T, Set<T>> _transitiveClosure<T extends Object>(
+  Map<T, List<T>> graph, {
+  bool Function(T, T)? equals,
+  int Function(T)? hashCode,
+  bool acyclic = false,
+}) {
+  if (equals != null) {
```
Comment (jakemac53) [inline test/transitive_closure_test.dart:336]: nit: consider asserting that `hashCode` and `equals` are either both null or both non-null. Or if we don't want to assert that then add `|| hashCode != null` here.

File: lib/src/strongly_connected_components.dart (hunk 2 above)
Comment (nex3) [reply lib/src/strongly_connected_components.dart]: It looks like I can actually just remove `late` and make this non-nullable and the compiler can tell it's being assigned.

**Review by jakemac53 — APPROVED** [review]
@natebosch did you want to look at this as well?

**Review by natebosch — APPROVED** [review]

**Pull request merged**